src/hotspot/share/runtime/threads.cpp

```diff
@@ -548,7 +548,7 @@ jint Threads::create_vm(JavaVMInitArgs* args, bool* canTryAgain) {
   // Initialize global modules
   jint status = init_globals();
   if (status != JNI_OK) {
-    main_thread->smr_delete();
+    // Don't delete main_thread here as we may need to report exceptions.
     *canTryAgain = false; // don't let caller call JNI_CreateJavaVM again
```

Comment:
So could the jni code try to create a vm, fail and then go on to do other stuff?  Would this then be a memory leak?

Reply:
maybe this should be a if (!HAS_PENDING_EXCEPTION) main_thread->smr_delete(); and then in the caller will exit with vm_exit_during_initialization() which does exit() so not necessary to delete the main thread.

Reply:
> Would this then be a memory leak?

Yes but at this point we are "leaking everything that VM initialization created but does not clean up. So adding a JavaThread instance to that seems insignificant.

> maybe this should be a if (!HAS_PENDING_EXCEPTION) main_thread->smr_delete(); 

That would work too. I can make that switch.